src/script/lua_api/l_areastore.cpp

```diff
@@ -0,0 +1,420 @@
+/*
+Minetest
+Copyright (C) 2015 est31 <mtest31@outlook.com>
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU Lesser General Public License as published by
+the Free Software Foundation; either version 2.1 of the License, or
+(at your option) any later version.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License along
+with this program; if not, write to the Free Software Foundation, Inc.,
+51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+
+#include "lua_api/l_areastore.h"
+#include "lua_api/l_internal.h"
+#include "common/c_converter.h"
+#include "cpp_api/s_security.h"
+#include "areastore.h"
+#include "filesys.h"
+#ifndef ANDROID
+#include "cmake_config.h"
+#endif
+#include <fstream>
+
+static inline void get_data_and_border_flags(lua_State *L, u8 start_i,
+		bool *borders, bool *data)
+{
+	if (!lua_isboolean(L, start_i))
+		return;
+	*borders = lua_toboolean(L, start_i);
+	if (!lua_isboolean(L, start_i + 1))
+		return;
+	*data = lua_toboolean(L, start_i + 1);
+}
+
+static inline void push_area(lua_State *L, const Area *a, bool borders, bool data)
+{
+	if (!(borders || data)) {
```

Comment:
You might think this is an optimization, because it's using two operators (! and ||) versus three (!, &&, !), but it is less clear and actually slightly slower than the more readable demorganed version!

This is because the logical 'not' operator instructs the compiler to insert a different version of the same instruction, rather than inserting an additional instruction.
Also, if you include both conditions in a parenthesized expression like that, it removes the opportunity to perform short circuit evaluation, so both conditions are always evaluated regardless of one's truth or falsehood.